mreg_cli/outputmanager.py

```diff
@@ -0,0 +1,128 @@
+"""This module contains the output management for the mreg_cli package.
+
+This is a singleton class that manages the output for the CLI. It stores the
+output lines and formats them for display. It also manages the filter for the
+command.
+"""
+
+import re
+from typing import Any, List, Tuple
+
+
+class OutputManager:
+    """Manages and formats output for display.
+
+    This is a singleton class to retain context between calls. It is initiated
+    and cleared in the main function of the CLI, all other calls should only
+    add lines to the output via add_line(), add_formatted_line(), or
+    add_formatted_line_with_source(),
+    """
+
+    _instance = None
+
+    def __new__(cls):
+        if cls._instance is None:
+            cls._instance = super(OutputManager, cls).__new__(cls)
+            cls._instance._lines = []
+        return cls._instance
+
+    def clear(self) -> None:
+        """Clears the object."""
+        self._lines = []
+        self._filter_re = None
+        self._negate = False
+        self._command = None
+
+    def has_output(self) -> bool:
+        """Returns True if there is output to display."""
+        return len(self._lines) > 0
+
+    def from_command(self, command: str) -> str:
+        """Adds the command that generated the output.
+
+        :param command: The command to add.
+
+        :return: The cleaned command, devoid of filters and other noise.
+        """
+        self.command = command
+        self._command, self._filter_re, self._negate = self.get_filter()
+        return self._command
+
+    def add_line(self, line: str) -> None:
+        """Adds a line to the output.
+
+        :param line: The line to add.
+        """
+        self._lines.append(line)
+
+    def add_formatted_line(self, key: str, value: str, padding: int = 14) -> None:
+        """Formats and adds a key-value pair as a line.
+
+        :param key: The key or label.
+        :param value: The value.
+        :param padding: The padding for the key.
+        """
+        formatted_line = "{1:<{0}} {2}".format(padding, key, value)
+        self.add_line(formatted_line)
+
+    def add_formatted_line_with_source(
+        self, key: str, value: str, source: str = "", padding: int = 14
+    ) -> None:
+        """Formats and adds a key-value pair as a line with a source.
+
+        :param key: The key or label.
+        :param value: The value.
+        :param source: The source of the value.
+        :param padding: The padding for the key.
+        """
+        formatted_line = "{1:<{0}} {2:<{0}} {3}".format(padding, key, value, source)
+        self.add_line(formatted_line)
+
+    # We want to use re.Pattern as the type here, but python 3.6 and older re-modules
+    # don't have that type. So we use Any instead.
+    def get_filter(self) -> Tuple[str, Any, bool]:
+        """Returns the filter for the output.
+
+        :return: The filter and whether it is a negated filter.
+        """
+        command = self.command
```

Comment:
What if the command was passed in as an argument instead, so that the method isn't reliant on `from_command()` setting the `self.command` attribute? Easier to test, etc.

Reply:
I have no recollection of why it ended up this way. It is much cleaner to pass things explicitly. I'll retain the self.command for now though, but assign it in get_filter instead. Thanks!